packages/frontend/src/features/sqlRunner/runners/SqlRunnerResultsRunnerFrontend.ts

```diff
@@ -0,0 +1,234 @@
+        }
+        return customMetric;
+    });
+    const groupBy = query.pivot?.on.map((on) => {
+        const f = fields.find((field) => field.name === on);
+        if (!f) {
+            throw new Error('Unexpected error: incorrect pivot configuration');
+        }
+        return f;
+    });
+
+    if (index === undefined || values === undefined || values.length === 0) {
+        throw new Error('Unexpected error: incorrect pivot configuration');
+    }
+    return {
+        indexColumn: {
+            reference: index.name,
+            type: getVizIndexTypeFromSemanticLayerFieldType(index.type),
+        },
+        valuesColumns: values.map((value) => ({
+            reference: value.name,
+            aggregation: value.aggType ?? VIZ_DEFAULT_AGGREGATION,
+        })),
+        groupByColumns: groupBy?.map((f) => ({ reference: f.name })),
+    };
+};
+
+// TEMPORARY
+export const getPivotQueryFunctionForSqlRunner = ({
+    projectUuid,
+    slug,
+    uuid,
+    limit,
+    sql,
+    fields,
+}: {
+    projectUuid: string;
+    slug?: string;
+    uuid?: string;
+    limit: number;
+    sql: string;
+    fields: SemanticLayerField[];
+}): RunPivotQuery => {
+    return async (query: SemanticLayerQuery) => {
+        const index = query.pivot?.index[0];
+        if (index === undefined) {
+            return {
+                results: [],
+                indexColumn: undefined,
+                valuesColumns: [],
+                columns: [],
+            };
+        }
+        const { indexColumn, valuesColumns, groupByColumns } =
+            convertSemanticLayerQueryToSqlRunnerPivotQuery(query, fields);
+        const pivotResults = await pivotQueryFn({
+            projectUuid,
+            slug,
+            uuid,
+            sql,
+            indexColumn,
+            valuesColumns,
+            groupByColumns,
+            limit,
+        });
+
+        const columns: VizColumn[] = [
+            ...(pivotResults.indexColumn?.reference
+                ? [pivotResults.indexColumn.reference]
+                : []),
+            ...pivotResults.valuesColumns,
+        ].map((field) => ({
+            reference: field,
+        }));
+
+        return {
+            results: pivotResults.results,
+            indexColumn: pivotResults.indexColumn,
+            valuesColumns: pivotResults.valuesColumns,
+            columns,
+        };
+    };
+};
+
+export class SqlRunnerResultsRunnerFrontend
+    extends BaseResultsRunner
+    implements IResultsRunner
+{
+    constructor({
+        columns,
+        rows,
+        projectUuid,
+        limit,
+        sql,
+        slug,
+        uuid,
+    }: {
+        columns: VizColumn[];
+        rows: RawResultRow[];
+        projectUuid: string;
+        limit: number;
+        sql: string;
+        slug?: string;
+        uuid?: string;
+    }) {
+        const fields: SemanticLayerField[] = columns.map((column) => ({
```

Comment:
Creates the fields - makes all the base logic shared for semantic viewer + sql runner